File: problems/scaler/1-D_Array/Rain_water_trapping.js
/**
 * Rain Water Trapping
 */

function getPfm(A) {
  const n = A.length;
  const pfm = new Array(n).fill(0);

  pfm[0] = A[0];

  for (let i = 1; i < n; i++) {
    pfm[i] = Math.max(pfm[i - 1], A[i]);
  }

  return pfm;
}

function getSfm(A) {
  const n = A.length;
  const sfm = new Array(n).fill(0);

  sfm[n - 1] = A[n - 1];

  for (let i = n - 2; i >= 0; i--) {
    sfm[i] = Math.max(sfm[i + 1], A[i]);
  }

  return sfm;
}

function soln(A) {
  const n = A.length;
  const pfm = getPfm(A);
  const sfm = getSfm(A);

  let water = 0;

  for (let i = 0; i < n; i++) {
    water += Math.min(pfm[i], sfm[i]) - A[i];
  }

  return water;
}

console.log(soln([2, 1, 3, 2, 1, 2, 4, 3, 2, 1, 3, 1])); // 8

File: src/main/java/com/dsa/problems/One_D_Array/Rain_water_trapping.java
/**
 * Rain Water Trapping
 */
package com.dsa.problems.One_D_Array;

import java.util.*;

public class Rain_water_trapping {

  public static ArrayList<Integer> getPfm(ArrayList<Integer> A) {
    int n = A.size();

    ArrayList<Integer> pfm = new ArrayList<>();

    for (int i = 0; i < n; i++) {
      pfm.add(0);
    }

    pfm.set(0, A.get(0));

    for (int i = 1; i < n; i++) {
      pfm.set(i, Math.max(pfm.get(i - 1), A.get(i)));
    }

    return pfm;
  }

  public static ArrayList<Integer> getSfm(ArrayList<Integer> A) {
    int n = A.size();

    ArrayList<Integer> sfm = new ArrayList<>();

    for (int i = 0; i < n; i++) {
      sfm.add(0);
    }

    sfm.set(n - 1, A.get(n - 1));

    for (int i = n - 2; i >= 0; i--) {
      sfm.set(i, Math.max(sfm.get(i + 1), A.get(i)));
    }

    return sfm;
  }

  public static int soln(ArrayList<Integer> A) {
    int n = A.size();

    ArrayList<Integer> pfm = getPfm(A);
    ArrayList<Integer> sfm = getSfm(A);

    int maxWater = 0;

    for (int i = 0; i < n; i++) {
      maxWater = maxWater + Math.min(pfm.get(i), sfm.get(i)) - A.get(i);
    }

    return maxWater;
  }

  public static void main(String[] args) {
    System.out.println(soln(new ArrayList<Integer>(Arrays.asList(2, 1, 3, 2, 1, 2, 4, 3, 2, 1, 3, 1))));
  }
}

// console.log(soln([2, 1, 3, 2, 1, 2, 4, 3, 2, 1, 3, 1])); // 8
